include/ethercat_sdk_master/EthercatActuatorDevice.h

```diff
@@ -0,0 +1,73 @@
+ **
+ ** 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
+ **
+ ** 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "EthercatDevice.hpp"
+
+namespace ecat_master {
+class EthercatActuatorDevice: public EthercatDevice
+{
+
+public:
+    typedef std::shared_ptr<EthercatActuatorDevice> SharedPtr;
+
+    class EthercatActuatorDeviceReading
+    {
+    public:
+        double currentQ = 0;        //A
+        double motorVelocity = 0;   //rad/s
+        double motorPosition = 0;   //rad
+        double temperature = 0;     //°C
+        std::chrono::system_clock::time_point timeStamp; //Time
+    };
+
+    enum class EthercatActuatorDeviceControlMode
+    {
+        None = 0,
+        Current = 1,
+        Velocity = 2,
+        Position = 3,
+    };
+
+    class EthercatActuatorDeviceCommand
+    {
+    public:
+        uint32_t mode = static_cast<uint32_t>(EthercatActuatorDeviceControlMode::None);
+        double currentQdes = 0; //A
+        double motorVelocityDes = 0; //rad/s
+        double motorPositionDes = 0; //rad
+    };
+
+    /**
+     * @brief getLatestReading
+     * @return The reading from the last update step
+     */
+    virtual std::shared_ptr<EthercatActuatorDeviceReading> getLatestReading() const = 0;
+    /**
+     * @brief setCommand - The command to be sent in the next update step
+     * @param command
+     * @return true in case of success
+     */
+    virtual bool setCommand(const std::shared_ptr<EthercatActuatorDeviceCommand>& command) = 0;
+
+    /**
+     * @brief getAvailableReadingFields
+     * @return vector of pairs that contain the name + a getter function in order to access a certain field in
+     */
+    virtual std::vector<std::pair<std::string, std::function<double(void)>>> getAvailableReadingFields(std::shared_ptr<EthercatActuatorDeviceReading> _reading) = 0;
```

Comment:
Okay, so you use the shared ptr to identify the reading, then this function can pass functions that would allow you to get the remaining fields. 

I would likely make this independent of the actuator reading since now for every reading, a vector of strings and functions have to be constructed and passed around.

Reply:
And why not return the values directly, instead of a function?